### PR TITLE
[11.0] web_export_view: Add filters in the 1st line of the exported file

### PR DESCRIPTION
**Pull request by jpcweb**
Add filters in the 1st line to the exported file

### CODE REVIEW

**Comment by pedrobaeza** [comment]
Next time please put a better title, not only `11.0`. I see that the changes seem to be higher than what you explain. Can you talk about this and put screenshots?

**Comment by jpcweb** [comment]
I'm sorry about the title, your are right.
The main purpose is to add the filters in the exported excel file.
Indeed it could be interested to have the context of the export when sharing the document.
I've attached a screenshot and the export file.

The .txt file is a .xls file (you just have to change the file extension to .xls)

filter_and_export

[product.template.txt](https://github.com/OCA/web/files/2583082/product.template.txt)

**Comment by pedrobaeza** [comment]
Thanks for explaining. I'm not sure this is so interesting, but I let other reviewers to decide.

File: web_export_view/controllers/controllers.py [hunk 1]
```diff
@@ -1,12 +1,34 @@
...
 # Copyright 2012 Therp BV
 # License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl.html).
 
-import json
 import odoo.http as http
+import io
+import babel.messages.pofile
+import base64
+import datetime
+import functools
```
Comment (moylop260) [inline web_export_view/controllers/controllers.py:13]: Could you remove the unused libraries, please?

Comment (moylop260) [reply web_export_view/controllers/controllers.py:13]: FYI you can check them from: https://travis-ci.org/OCA/web/jobs/455191769#L564

**Review by moylop260 — COMMENTED** [review]
Is possible add a unittest?

File: web_export_view/controllers/controllers.py [hunk 2]
```diff
@@ -16,15 +38,72 @@ def __getattribute__(self, name):
...
+
+        base_style = xlwt.easyxf('align: wrap yes')
+        date_style = xlwt.easyxf('align: wrap yes', num_format_str='YYYY-MM-DD')
+        datetime_style = xlwt.easyxf('align: wrap yes', num_format_str='YYYY-MM-DD HH:mm:SS')
+
+        for row_index, row in enumerate(rows):
+            for cell_index, cell_value in enumerate(row):
+                cell_style = base_style
+
+                if isinstance(cell_value, bytes) and not isinstance(cell_value, pycompat.string_types):
```
Comment (simahawk) [inline web_export_view/controllers/controllers.py:72]: I'd move this block out to a method returning cell_value and cell_style

File: web_export_view/controllers/controllers.py [hunk 3]
```diff
@@ -16,15 +38,72 @@ def __getattribute__(self, name):
...
+                    cell_value = re.sub("\r", " ", pycompat.to_text(cell_value))
+                    # Excel supports a maximum of 32767 characters in each cell:
+                    cell_value = cell_value[:32767]
+                elif isinstance(cell_value, datetime.datetime):
+                    cell_style = datetime_style
+                elif isinstance(cell_value, datetime.date):
+                    cell_style = date_style
+                worksheet.write(row_index + pos + 1, cell_index, cell_value, cell_style)
+
+        fp = io.BytesIO()
```
Comment (simahawk) [inline web_export_view/controllers/controllers.py]: use a context manager

File: web_export_view/controllers/controllers.py [hunk 4]
```diff
@@ -16,15 +38,72 @@ def __getattribute__(self, name):
...
     def export_xls_view(self, data, token):
         data = json.loads(data)
         model = data.get('model', [])
+        columns_searchs = data.get('searchs', [])
         columns_headers = data.get('headers', [])
         rows = data.get('rows', [])
 
         return request.make_response(
-            self.from_data(columns_headers, rows),
+            self.from_data(columns_headers, rows, columns_searchs),
```
Comment (simahawk) [inline web_export_view/controllers/controllers.py:108]: `search=columns_search`

File: web_export_view/static/src/js/web_export_view.js [hunk 5]
```diff
@@ -42,10 +42,16 @@ odoo.define('web_export_view', function (require) {
                     return true;
                 });
             }
+            var export_columns_searchs = [];
             var export_columns_keys = [];
             var export_columns_names = [];
             var column_index = 0;
             var column_header_selector;
+
+            $('.o_facet_values > span').each(function() {
```
Comment (simahawk) [inline web_export_view/static/src/js/web_export_view.js:51]: `this.$('.o_facet_values > span')`

**Review by simahawk — CHANGES_REQUESTED** [review]
please fix jslint + pylint.
Also, a screenshot of the final result to add to README would be nice.

File: web_export_view/controllers/controllers.py [hunk 6]
```diff
@@ -1,30 +1,111 @@
...
+        if len(rows) > excl_limit:
+            print(_('There are too many rows (%s rows, limit: 65535) to export as Excel 97-2003 (.xls) format. Consider splitting the export.') % len(rows))
+
+        workbook = xlsxwriter.Workbook(output, {'in_memory': True})
+        worksheet = workbook.add_worksheet('Sheet 1')
+
+        pos = 0
+
+        if len(searchs) > 0 :
+            worksheet.write(pos, 0, "Les filtres: ")
```
Comment (tarteo) [inline web_export_view/controllers/controllers.py:54]: Please use english `_t('Les filtres')`

**Review by tarteo — CHANGES_REQUESTED** [review]

**Comment by MiquelRForgeFlow** [comment]
conflicts

**Review by daramousk — CHANGES_REQUESTED** [review]
Hello and thanks for this,

Now, since this has been inactive for a while, I would like to ask if @jpcweb will continue with this or should someone else take over, if yes then please do the following so that people can properly review

1) Make everything in CI pipeline green, check the error messages one by one and fix them
2) Add tests. This seems like a complicated implementation and we need to be sure nothing breaks
3) Rebase on top of 11.0 to fix the conflicts.

Then we can continue with reviewing.

Thanks

**Pull request closed without merging**